src/Build/Utilities/EngineFileUtilities.cs

```diff
@@ -154,11 +191,109 @@ FileMatcher fileMatcher
+                            }
+                            else if ((!string.IsNullOrWhiteSpace(excludeFileSpec)) || (includeLocation == null)) // Latter condition necessary for GetFileListUnescaped call in ItemGroupIntrinsicTask to get caught in here.
+                            {
+                                if (includeLocation == null)
+                                {
+                                    ProjectErrorUtilities.ThrowInvalidProject(excludeLocation, "WildcardResultsInDriveEnumeration", EscapingUtilities.UnescapeAll(filespecUnescaped), XMakeAttributes.exclude, XMakeElements.itemGroup);
+                                }
+                                else
+                                {
+                                    ProjectErrorUtilities.ThrowInvalidProject(excludeLocation, "WildcardResultsInDriveEnumeration", EscapingUtilities.UnescapeAll(excludeFileSpec), XMakeAttributes.exclude, XMakeElements.itemGroup);
+                                }
+                            }
+
+                            break;
+
+                        case ILoggingService loggingService:  // Logging mechanism used for Evaluator
+                            ProjectErrorUtilities.ThrowInvalidProject(includeLocation, "WildcardResultsInDriveEnumeration", EscapingUtilities.UnescapeAll(filespecUnescaped), XMakeAttributes.include, XMakeElements.itemGroup);
+                            break;
+
+                        case EvaluationLoggingContext evaluationLoggingContext:  // Logging mechanism used for Evaluator and LazyItemEvaluator.IncludeOperation
+                            if (importLocation != null)
+                            {
+                                ProjectErrorUtilities.ThrowInvalidProject(importLocation, "WildcardResultsInDriveEnumeration", EscapingUtilities.UnescapeAll(filespecUnescaped), XMakeAttributes.project, XMakeElements.import);
+                            }
+                            else if (string.IsNullOrWhiteSpace(excludeFileSpec))
+                            {
+                                ProjectErrorUtilities.ThrowInvalidProject(includeLocation, "WildcardResultsInDriveEnumeration", EscapingUtilities.UnescapeAll(filespecUnescaped), XMakeAttributes.include, XMakeElements.itemGroup);
+                            }
+                            else
+                            {
+                                ProjectErrorUtilities.ThrowInvalidProject(excludeLocation, "WildcardResultsInDriveEnumeration", EscapingUtilities.UnescapeAll(excludeFileSpec), XMakeAttributes.exclude, XMakeElements.itemGroup);
+                            }
+
+                            break;
+
+                        default:
+                            throw new NotImplementedException();
```

Comment:
A message like this would smooth the path to fixing a problem here:

```suggestion
                            throw new NotImplementedException($"Logging type {loggingMechanism.GetType()} is not understood by {nameof(GetFileList)}.");
```